Resolve named ports for DNS policies

jrajahalme opened the pull request:
Implement `l4Policy` wrapper with resolved named port so that proxy package can get the actual resolved named port number when calling `GetPort`. Previously, if policy used a named port, `GetPort` returned `0`.

`GetPort` is only used for DNS proxy function `UpdateAllowed`. This means that using a named port for a DNS policy destination port likely has not functioned as intended. Since DNS port is typically `53`, it seems likely that named ports would not be used in DNS policies in practice.

Fixes: #11092

```release-note
Named ports in DNS policies are now resolved correctly.
```

jrajahalme committed:
endpoint: Make redirect_test more robust

Adding more tests fails due to missing selector cache notifier mock and
leaking of endpoint identities in a global manager between tests. Address
these and also remove unnecessary logging.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aditighag reviewed (commented):
I don't have much context. Perhaps it's best reviewed by someone from the policy team. /cc @cilium/sig-policy 

>Since DNS port is typically 53, it seems likely that named ports would not be used in DNS policies in practice.

I don't quite follow this. K8s mandates named ports when there are more than one port defined in a pod spec.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -172,6 +172,45 @@ func (e *Endpoint) writeHeaderfile(prefix string) error {
+func (p *proxyPolicy) GetListener() string {
+	return p.l4.GetListener()
+}
+
```
We can simplify a little bit by only implementing the methods that are different.
```suggestion
type proxyPolicy struct {
	*policy.L4Filter
	port uint16
}

// newProxyPolicy returns a new instance of proxyPolicy by value
func (e *Endpoint) newProxyPolicy(l4 *policy.L4Filter, port uint16) proxyPolicy {
	return proxyPolicy{L4Filter: l4, port: port}
}

// GetPort returns the destination port number on which the proxy policy applies
// This version properly returns the port resolved from a named port, if any.
func (p *proxyPolicy) GetPort() uint16 {
	return p.port
}
```

jrajahalme replied:
Done, thanks!

aanm reviewed (approved):
Just one tiny suggestion, I'm fine either case.

christarazi reviewed (approved):
LGTM, I agree with @aanm's suggestion which simplifies the code.

jrajahalme committed:
endpoint: Resolve named ports for redirects

Implement l4Policy wrapper with resolved named port so that proxy package
can get the actual resolved named port number when calling
GetPort(). Previously, if policy used a named port, GetPort returned 0.

GetPort() is only used for DNS proxy function UpdateAllowed(). This means
that using a named port for a DNS policy destination port likely has not
functioned as intended. Since DNS port is typically 53, it seems likely
that named ports are not used in DNS policies.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
